### DISCUSSION TITLE
Is it possible to use CTest?

### DISCUSSION THREAD

**Discussion by ylz-at** [android/ndk#1726] [Q&A] [answered]
I managed to write some parts of my App in C++(building with CMake) with the help of NDK.
Normally on Windows/Mac, I use GTest to write test cases and run `ctest` to run the tests.
But I don't know how to test the C++ code on Android.

What is the best approach for running CMake tests on Android?

**Comment by jfgoog** [comment, marked as answer]
If you want to run googletest on-device, take a look at https://github.com/jfgoog/ndk-examples/tree/main/googletest

**Reply by DanAlbert** [reply]
(We'll get the various forms of that sample turned into a proper sample and doc once it's a bit more production ready. There are some quirks as you can see in `packagingOptions`)

**Reply by ylz-at** [reply]
I found out that test executables can directly run on Android.
1. cross compile test executables(`add_executable` instead of add_library) using cmake by following [the guide](https://developer.android.com/ndk/guides/cmake)
2. adb push to /data/local/tmp/nativetest
3. adb shell "chmod +x /data/local/tmp/nativetest/hello"
4. adb shell "/data/local/tmp/nativetest/hello"

There is a CMake utility `AndroidTestUtilities` that can push things to android devices, but I didn't figure out how to use it to push multiple executables(can't find any example). If the utility works, then it will be easy to integrate with CMake with `add_test adb shell /data/local/tmp/nativetest/${TEST_NAME}`.

The JNI+junit approach might work well if you have Java/Kotlin in the application, but the toy app I am working on is pure C++ and we made `android:hasCode="false"`.

**Reply by jfgoog** [reply]
Yes, that is possible. In fact, we do it for some of our tests of the NDK itself.

But we don't recommend it for app developers, because code executed with "adb shell" runs in a different, more permissive environment than an Android app. For example, if the test code accesses the internet, the proper permission must be set in the app.

**Comment by triplef** [comment]
I was wondering the same in order to test our library and I found the following CMake module that seems to allow running CMake tests on Android devices/emulators:
https://gitlab.xiph.org/xiph/opus/-/merge_requests/28/diffs

There’s also this official module, but I’m not sure if that helps for this:
https://cmake.org/cmake/help/latest/module/AndroidTestUtilities.html
